tracee-ebpf/tracee/tracee.bpf.c

```diff
@@ -2064,6 +2169,309 @@ int BPF_KPROBE(trace_cap_capable)
+
+    // getting struct sock from socket
+    bpf_probe_read(&sk, sizeof(sk), &sock->sk);
+
+    // getting socket family
+    bpf_probe_read(&family, sizeof(family), &sk->sk_family);
+    if ( (family != AF_INET) && (family != AF_INET6) ) {
+        return 0;
+    }
+
+    context_t context = init_and_save_context(ctx, submit_p, SECURITY_SOCKET_LISTEN, 2 /*argnum*/, 0 /*ret*/);
+
+    // getting event tags
+    u64 *tags = bpf_map_lookup_elem(&params_names_map, &context.eventid);
+    if (!tags) {
+        return -1;
+    }
+
+    if ( family == AF_INET ){
+
+        net_conn_v4_t net_details = {};
+
+        get_network_details_from_sock_v4(sk, &net_details, 0);
+
+        struct sockaddr_in local;
+        local.sin_family = family;
+        local.sin_port = net_details.local_port;
+        struct in_addr local_addr;
+        local_addr.s_addr = net_details.local_address;
+        local.sin_addr = local_addr;
+
+        save_to_submit_buf(submit_p, (void *)&local, sizeof(struct sockaddr_in), SOCKADDR_T, DEC_ARG(0, *tags));
+
+    }
+    else if ( family == AF_INET6 ){
+        net_conn_v6_t net_details = {};
+
+        get_network_details_from_sock_v6(sk, &net_details, 0);
+
+        struct sockaddr_in6 local;
+        local.sin6_family = family;
+        local.sin6_port = net_details.local_port;
+        local.sin6_flowinfo = 0;
+        local.sin6_addr = net_details.local_address;
+        local.sin6_scope_id = 0;
+
+        save_to_submit_buf(submit_p, (void *)&local, sizeof(struct sockaddr_in6), SOCKADDR_T, DEC_ARG(0, *tags));
+    }
+
+    save_to_submit_buf(submit_p, (void *)&backlog, sizeof(int), INT_T, DEC_ARG(1, *tags));
+
+    events_perf_submit(ctx);
+    return 0;
+};
+
+SEC("kprobe/security_socket_connect")
+int BPF_KPROBE(trace_security_socket_connect)
+{
+
+    // trace the event security_socket_connect
+
+    if (!should_trace())
+        return 0;
+
+    struct socket *sock = (struct socket *)PT_REGS_PARM1(ctx);
+
+    buf_t *submit_p = get_buf(SUBMIT_BUF_IDX);
+    if (submit_p == NULL)
+        return 0;
+
+    set_buf_off(SUBMIT_BUF_IDX, sizeof(context_t));
+
+    context_t context = init_and_save_context(ctx, submit_p, SECURITY_SOCKET_CONNECT, 1 /*argnum*/, 0 /*ret*/);
+
+    // getting event tags
+    u64 *tags = bpf_map_lookup_elem(&params_names_map, &context.eventid);
+    if (!tags) {
+        return -1;
+    }
+
+    // getting source details
+
+    u16 family;
+    struct sock *sk;
+
+    // getting struct sock from socket
+    bpf_probe_read(&sk, sizeof(sk), &sock->sk);
+
+    // getting socket family
+    bpf_probe_read(&family, sizeof(family), &sk->sk_family);
+    if ( (family != AF_INET) && (family != AF_INET6) ) {
+        return 0;
+    }
+
+    // getting destination details
+
+    struct sockaddr *address = (struct sockaddr *)PT_REGS_PARM2(ctx);
+    sa_family_t sa_fam;
+
+    // read family from struct sockaddr*
+    bpf_probe_read(&sa_fam, sizeof(sa_fam), &address->sa_family);
```

Comment:
You read family here and from sk->sk_family in 2315.
Any reason for the value to be different between these two?

Reply:
these values come from the user - the user creates the socket with a specified family, and the user decides which address to connect to as well as the family of the address. so these values can be different.
the connect syscall checks if the two families are aligned, if not it returns an error EAFNOSUPPORT. but this happens after the LSM hook is called.

Reply:
In that case, I think it is enough to use the family given by the user's address.
We are more interested in the intentions of the user here, and if there is no match, we can see it in the return value from the syscall. This will allow us to remove some lines here and save some cycles